src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -583,6 +583,75 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id, HR
     }
 
+#ifndef _WIN32
+    // Fix PInvokeMap (Non windows only)
+    if (module_info.assembly.name == managed_profiler_name)
+    {
+        Info("ModuleLoadFinished: ", managed_profiler_name," - Fix PInvoke maps");
+
+        // We are in the right module, so we try to load the mdTypeDef from the target type name.
+        mdTypeDef nativeMethodsTypeDef = mdTypeDefNil;
+        auto foundType = FindTypeDefByName(nonwindows_nativemethods_type,
+                                           module_metadata->assemblyName, metadata_import, nativeMethodsTypeDef);
+        if (foundType)
+        {
+            // Define the actual profiler file path as a ModuleRef
+            WSTRING native_profiler_file = GetProfilerFilePath();
+            mdModuleRef profiler_ref;
+            hr = metadata_emit->DefineModuleRef(native_profiler_file.c_str(), &profiler_ref);
+            if (SUCCEEDED(hr))
+            {
+                // Enumerate all methods inside the native methods type with the PInvokes
+                Enumerator<mdMethodDef> enumMethods = Enumerator<mdMethodDef>(
+                    [metadata_import, nativeMethodsTypeDef](HCORENUM* ptr, mdMethodDef arr[], ULONG max, ULONG* cnt) -> HRESULT
+                    {
+                        return metadata_import->EnumMethods(ptr, nativeMethodsTypeDef, arr, max, cnt);
+                    }, [metadata_import](HCORENUM ptr) -> void { metadata_import->CloseEnum(ptr); });
+
+                EnumeratorIterator<mdMethodDef> enumIterator = enumMethods.begin();
+                while (enumIterator != enumMethods.end())
+                {
+                    auto methodDef = *enumIterator;
+
+                    // Get the current PInvoke map to extract the flags and the entrypoint name
+                    DWORD pdwMappingFlags;
+                    WCHAR importName[kNameMaxSize]{};
+                    DWORD importNameLength = 0;
+                    mdModuleRef importModule;
+                    hr = metadata_import->GetPinvokeMap(methodDef, &pdwMappingFlags, importName, kNameMaxSize,
+                                                        &importNameLength, &importModule);
+                    if (SUCCEEDED(hr))
+                    {
+                        // Delete the current PInvoke map
+                        hr = metadata_emit->DeletePinvokeMap(methodDef);
+                        if (SUCCEEDED(hr))
+                        {
+                            // Define a new PInvoke map with the new ModuleRef of the actual profiler file path
+                            hr = metadata_emit->DefinePinvokeMap(methodDef, pdwMappingFlags,
+                                                                 WSTRING(importName).c_str(), profiler_ref);
+                            if (FAILED(hr))
+                            {
+                                Warn("ModuleLoadFinished: DefinePinvokeMap failed");
```

Comment:
As part of the failure path, should we attempt to re-define the previous PinvokeMap?

Reply:
That makes sense... will do the change.

Reply:
done in b053eb8